File: integrations/claude-code/scripts/idle-watcher.py
#!/usr/bin/env python3
"""Idle watcher daemon — persists quiet sessions into Cognee.

Launched detached from ``session-start.py``. Polls
``~/.cognee-plugin/activity.ts`` every ``POLL_SECONDS``. When the last
activity is older than ``IDLE_SECONDS`` and we haven't bridged since
that point, persists the session cache and refreshes graph context.

Stops cleanly on:
  * ``~/.cognee-plugin/watcher.stop`` sentinel file.
  * Receiving SIGTERM (from SessionEnd hook or manual kill).
  * The pidfile being overwritten by a newer watcher (restart case).

Survives SessionEnd / Claude crashes better than the SessionEnd hook
does — that hook won't run if Claude was killed hard.
"""

import asyncio
import json
import os
import signal
import sys
import time
from contextlib import nullcontext
from pathlib import Path
from typing import Optional

# Tunable via env. Defaults chosen to avoid thrashing the LLM: 60s idle
# threshold means you have to actively pause a full minute, and the 20s
# bridge cooldown prevents back-to-back runs when activity is sporadic.
POLL_SECONDS = float(os.environ.get("COGNEE_IDLE_POLL", "10"))
IDLE_SECONDS = float(os.environ.get("COGNEE_IDLE_THRESHOLD", "60"))
IMPROVE_COOLDOWN = float(os.environ.get("COGNEE_IMPROVE_COOLDOWN", "120"))

_PLUGIN_DIR = Path.home() / ".cognee-plugin"
_ACTIVITY = _PLUGIN_DIR / "activity.ts"
_PIDFILE = _PLUGIN_DIR / "watcher.pid"
_STOPFILE = _PLUGIN_DIR / "watcher.stop"
_LOGFILE = _PLUGIN_DIR / "watcher.log"

# Script-local stop flag flipped by SIGTERM handler.
_should_stop = False


def _log(event: str, **detail) -> None:
    try:
        _PLUGIN_DIR.mkdir(parents=True, exist_ok=True)
        line = {"ts": time.time(), "pid": os.getpid(), "event": event}
        if detail:
            line["detail"] = detail
        with _LOGFILE.open("a", encoding="utf-8") as fh:
            fh.write(json.dumps(line, default=str) + "\n")
    except Exception:
        pass


def _read_activity_ts() -> Optional[float]:
    if not _ACTIVITY.exists():
        return None
    try:
        return float(_ACTIVITY.read_text(encoding="utf-8").strip())
    except Exception:
        return None


def _owns_pidfile() -> bool:
    """Return True if the pidfile still points at us."""
    try:
        return int(_PIDFILE.read_text(encoding="utf-8").strip()) == os.getpid()
    except Exception:
        return False


def _install_signal_handlers() -> None:
    def _handler(signum, _frame):
        global _should_stop
        _should_stop = True
        _log("signal_received", signum=signum)

    signal.signal(signal.SIGTERM, _handler)
    signal.signal(signal.SIGINT, _handler)


async def _improve_once(session_id: str, dataset: str, config: dict) -> bool:
    """Fire one session bridge cycle. Returns True on success."""
    sys.path.insert(0, os.path.dirname(__file__))
    try:
        from _plugin_common import (  # type: ignore
            persist_session_cache_to_graph_via_http,
            sync_lock,
        )

        lock = sync_lock("idle-watcher")
    except Exception as exc:
        _log("sync_lock_import_error", error=str(exc)[:200])
        lock = nullcontext(True)

    with lock as acquired:
        if not acquired:
            _log("bridge_skipped_lock_busy", session=session_id, dataset=dataset)
            return False

        try:
            from config import (  # type: ignore
                ensure_cognee_ready,
                ensure_dataset_ready,
                ensure_identity,
                is_cloud_mode,
                persist_session_cache_to_graph,
                sync_graph_context_to_session,
            )

            if is_cloud_mode(config):
                wrote = persist_session_cache_to_graph_via_http(dataset, session_id)
                _log(
                    "session_bridge_done",
                    session=session_id,
                    dataset=dataset,
                    via="http_remember",
                    wrote=wrote,
                )
                return True

            await ensure_cognee_ready(config)
            user_id, _ = await ensure_identity(config)

            from uuid import UUID

            from cognee.modules.users.methods import get_user

            user = await get_user(UUID(user_id)) if user_id else None
            if user:
                await ensure_dataset_ready(dataset, user)
                wrote = await persist_session_cache_to_graph(dataset, session_id, user)
                graph_result = await sync_graph_context_to_session(dataset, session_id, user)
                _log(
                    "session_bridge_done",
                    session=session_id,
                    dataset=dataset,
                    wrote=wrote,
                    graph_synced=graph_result.get("synced", 0),
                )
            return True
        except Exception as exc:
            _log("bridge_error", error=str(exc)[:300])
            return False


async def _main_loop(session_id: str, dataset: str, config: dict) -> None:
    _log("started", session=session_id, dataset=dataset, poll=POLL_SECONDS, idle=IDLE_SECONDS)
    last_improved_at = 0.0
    exit_reason = "loop_complete"

    while not _should_stop:
        if _STOPFILE.exists():
            _log("stop_sentinel_seen")
            exit_reason = "stop_sentinel"
            break
        if not _owns_pidfile():
            _log("pidfile_replaced")
            exit_reason = "pidfile_replaced"
            break

        now = time.time()
        ts = _read_activity_ts()
        if ts is None:
            await asyncio.sleep(POLL_SECONDS)
            continue

        idle_for = now - ts
        time_since_improve = now - last_improved_at
        if idle_for >= IDLE_SECONDS and time_since_improve >= IMPROVE_COOLDOWN:
            _log("idle_trigger", idle_for=round(idle_for, 1))
            ok = await _improve_once(session_id, dataset, config)
            if ok:
                last_improved_at = time.time()
                _log("bridge_done")
                exit_reason = "bridge_complete"
                break

        await asyncio.sleep(POLL_SECONDS)

    if _should_stop:
        exit_reason = "signal"

    ts = _read_activity_ts()
    if exit_reason in {"signal", "stop_sentinel"} and ts and ts > last_improved_at:
        _log("shutdown_trigger", reason=exit_reason, activity_age=round(time.time() - ts, 1))
        ok = await _improve_once(session_id, dataset, config)
        if ok:
            last_improved_at = time.time()
            _log("shutdown_bridge_done")
        else:
            _log("shutdown_bridge_failed")

    _log("exiting", reason=exit_reason)
    try:
        if _owns_pidfile():
            _PIDFILE.unlink()
    except Exception:
        pass


def main():
    _PLUGIN_DIR.mkdir(parents=True, exist_ok=True)

    # Config passed as a single JSON arg to avoid shell-quoting hazards.
    if len(sys.argv) < 2:
        _log("fatal_missing_args")
        sys.exit(1)
    try:
        bootstrap = json.loads(sys.argv[1])
    except Exception as exc:
        _log("fatal_bad_args", error=str(exc)[:200])
        sys.exit(1)

    session_id = bootstrap.get("session_id", "")
    dataset = bootstrap.get("dataset", "claude_sessions")
    try:
        from config import load_config  # type: ignore

        config = load_config()
        config.update({k: v for k, v in bootstrap.get("config", {}).items() if v})
    except Exception:
        config = bootstrap.get("config", {})
    if not session_id:
        _log("fatal_no_session_id")
        sys.exit(1)

    try:
        _PIDFILE.write_text(str(os.getpid()), encoding="utf-8")
    except Exception as exc:
        _log("pidfile_write_failed", error=str(exc)[:200])
        sys.exit(1)

    # Make sure a stale stop sentinel from a prior run doesn't kill us
    # the moment we start.
    try:
        if _STOPFILE.exists():
            _STOPFILE.unlink()
    except Exception:
        pass

    _install_signal_handlers()

    try:
        asyncio.run(_main_loop(session_id, dataset, config))
    except Exception as exc:
        _log("fatal_loop_error", error=str(exc)[:300])


if __name__ == "__main__":
    main()

File: integrations/vellum-assistant/src/idle-watcher.ts
/**
 * Idle watcher daemon — persists quiet sessions into Cognee.
 *
 * Ported from idle-watcher.py. A background daemon that:
 *   - Polls the activity file timestamp
 *   - When idle for > IDLE_SECONDS, triggers sync-session-to-graph
 *   - Stops on SIGTERM or watcher.stop sentinel file
 *
 * Launched detached from session-start. Polls
 * `~/.cognee-plugin/vellum-assistant/activity.ts` every POLL_SECONDS.
 * When the last activity is older than IDLE_SECONDS and we haven't
 * bridged since that point, persists the session cache to the graph.
 */

import { existsSync, readFileSync, writeFileSync, unlinkSync, mkdirSync } from "node:fs";
import { join } from "node:path";
import { spawn } from "bun";

import { hookLog, pluginStateDir, loadConfig } from "./plugin-common.ts";

// ─── Tunables (via env) ───────────────────────────────────────────────────────

const POLL_SECONDS = Number(process.env.COGNEE_IDLE_POLL ?? 10);
const IDLE_SECONDS = Number(process.env.COGNEE_IDLE_THRESHOLD ?? 60);
const IMPROVE_COOLDOWN = Number(process.env.COGNEE_IMPROVE_COOLDOWN ?? 120);

// ─── Paths ────────────────────────────────────────────────────────────────────

function activityPath(): string {
  return join(pluginStateDir(), "activity.ts");
}

function pidfilePath(): string {
  return join(pluginStateDir(), "watcher.pid");
}

function stopfilePath(): string {
  return join(pluginStateDir(), "watcher.stop");
}

function logPath(): string {
  return join(pluginStateDir(), "watcher.log");
}

// ─── State ────────────────────────────────────────────────────────────────────

let shouldStop = false;

// ─── Helpers ──────────────────────────────────────────────────────────────────

function readActivityTs(): number | null {
  if (!existsSync(activityPath())) return null;
  try {
    return parseFloat(readFileSync(activityPath(), "utf-8").trim());
  } catch (err) {
    hookLog("idle_watcher_activity_read_failed", { error: String(err).slice(0, 200) });
    return null;
  }
}

function ownsPidfile(): boolean {
  try {
    return parseInt(readFileSync(pidfilePath(), "utf-8").trim(), 10) === process.pid;
  } catch {
    return false;
  }
}

function installSignalHandlers(): void {
  const handler = (sig: string) => {
    shouldStop = true;
    hookLog("idle_watcher_signal_received", { signal: sig });
  };

  process.on("SIGTERM", () => handler("SIGTERM"));
  process.on("SIGINT", () => handler("SIGINT"));
}

// ─── Bridge (sync to graph) ───────────────────────────────────────────────────

/**
 * Fire one session bridge cycle. Spawns the sync-session-to-graph module
 * as a detached process. Returns true on success.
 */
async function improveOnce(
  sessionId: string,
  dataset: string,
): Promise<boolean> {
  try {
    // Import the sync module and call syncSessionToGraph directly.
    // This keeps everything in-process — no subprocess needed.
    const { syncSessionToGraph } = await import("./sync-session-to-graph.ts");
    const ok = await syncSessionToGraph(false);
    hookLog("idle_watcher_session_bridge_done", {
      session: sessionId,
      dataset,
      via: "http_remember",
      ok,
    });
    return ok;
  } catch (err) {
    hookLog("idle_watcher_bridge_error", { error: String(err).slice(0, 300) });
    return false;
  }
}

// ─── Main loop ────────────────────────────────────────────────────────────────

export interface IdleWatcherConfig {
  session_id: string;
  dataset: string;
  user_id?: string;
  session_key?: string;
  config?: Record<string, unknown>;
}

/**
 * Run the idle watcher main loop.
 *
 * @param config Bootstrap config from the launching hook.
 */
export async function runIdleWatcher(config: IdleWatcherConfig): Promise<void> {
  const sessionId = config.session_id || "";
  const dataset = config.dataset || "agent_sessions";

  if (!sessionId) {
    hookLog("idle_watcher_fatal_no_session_id");
    return;
  }

  // Ensure the plugin state dir exists.
  try {
    mkdirSync(pluginStateDir(), { recursive: true });
  } catch {
    // Best-effort.
  }

  // Write the pidfile.
  try {
    writeFileSync(pidfilePath(), String(process.pid), "utf-8");
  } catch (err) {
    hookLog("idle_watcher_pidfile_write_failed", { error: String(err).slice(0, 200) });
    return;
  }

  // Clear any stale stop sentinel from a prior run.
  try {
    if (existsSync(stopfilePath())) {
      unlinkSync(stopfilePath());
    }
  } catch (err) {
    hookLog("idle_watcher_stopfile_unlink_failed", { error: String(err).slice(0, 200) });
  }

  // Set env vars for the sync module.
  if (config.session_key) {
    process.env.COGNEE_SESSION_KEY = config.session_key;
  }

  installSignalHandlers();

  hookLog("idle_watcher_started", {
    session: sessionId,
    dataset,
    user_id: config.user_id ?? "",
    poll: POLL_SECONDS,
    idle: IDLE_SECONDS,
  });

  let lastImprovedAt = 0.0;
  let exitReason = "loop_complete";
  let bridgeDisabled = false;

  while (!shouldStop) {
    // Check stop conditions.
    if (existsSync(stopfilePath())) {
      hookLog("idle_watcher_stop_sentinel_seen");
      exitReason = "stop_sentinel";
      break;
    }
    if (!ownsPidfile()) {
      hookLog("idle_watcher_pidfile_replaced");
      exitReason = "pidfile_replaced";
      break;
    }

    const now = Date.now() / 1000;
    const ts = readActivityTs();
    if (ts === null) {
      await sleep(POLL_SECONDS * 1000);
      continue;
    }

    const idleFor = now - ts;
    const timeSinceImprove = now - lastImprovedAt;

    if (
      !bridgeDisabled &&
      idleFor >= IDLE_SECONDS &&
      timeSinceImprove >= IMPROVE_COOLDOWN
    ) {
      hookLog("idle_watcher_idle_trigger", { idle_for: Math.round(idleFor * 10) / 10 });
      const ok = await improveOnce(sessionId, dataset);
      if (ok) {
        lastImprovedAt = Date.now() / 1000;
        hookLog("idle_watcher_bridge_done");
        exitReason = "bridge_complete";
        break;
      }
      bridgeDisabled = true;
      hookLog("idle_watcher_bridge_disabled_after_failure");
    }

    await sleep(POLL_SECONDS * 1000);
  }

  if (shouldStop) {
    exitReason = "signal";
  }

  // On shutdown (signal or stop sentinel), do one final bridge if there's
  // been activity since the last improve.
  const ts = readActivityTs();
  if (
    !bridgeDisabled &&
    (exitReason === "signal" || exitReason === "stop_sentinel") &&
    ts !== null &&
    ts > lastImprovedAt
  ) {
    hookLog("idle_watcher_shutdown_trigger", {
      reason: exitReason,
      activity_age: Math.round((Date.now() / 1000 - ts) * 10) / 10,
    });
    const ok = await improveOnce(sessionId, dataset);
    if (ok) {
      lastImprovedAt = Date.now() / 1000;
      hookLog("idle_watcher_shutdown_bridge_done");
    } else {
      hookLog("idle_watcher_shutdown_bridge_failed");
    }
  }

  hookLog("idle_watcher_exiting", { reason: exitReason });

  // Clean up the pidfile.
  try {
    if (ownsPidfile()) {
      unlinkSync(pidfilePath());
    }
  } catch (err) {
    hookLog("idle_watcher_pidfile_unlink_failed", { error: String(err).slice(0, 200) });
  }
}

// ─── Launcher ─────────────────────────────────────────────────────────────────

/**
 * Launch the idle watcher as a detached background process.
 *
 * Called from session-start. The watcher re-execs this module file with
 * the config JSON as an argument.
 *
 * @param config The watcher configuration.
 * @returns The spawned process PID, or 0 on failure.
 */
export function launchIdleWatcher(config: IdleWatcherConfig): number {
  // If a watcher is already alive, kill it so the new one takes over.
  if (existsSync(pidfilePath())) {
    try {
      const pid = parseInt(readFileSync(pidfilePath(), "utf-8").trim(), 10);
      if (pid > 1 && pid !== process.pid) {
        try {
          process.kill(pid, "SIGTERM");
        } catch {
          // Already dead.
        }
      }
    } catch {
      // Best-effort.
    }
  }

  // Clear any stale stop sentinel.
  try {
    if (existsSync(stopfilePath())) {
      unlinkSync(stopfilePath());
    }
  } catch {
    // Best-effort.
  }

  try {
    const env: Record<string, string> = { ...process.env } as Record<string, string>;
    if (config.session_key) {
      env.COGNEE_SESSION_KEY = config.session_key;
    }

    const modulePath = new URL(import.meta.url).pathname;
    const proc = spawn({
      cmd: ["bun", "run", modulePath, JSON.stringify(config)],
      stdin: "null",
      stdout: logPath(),
      stderr: logPath(),
      env,
      detached: true,
    });

    hookLog("idle_watcher_launched", {
      session: config.session_id,
      dataset: config.dataset,
      pid: proc.pid,
    });

    return proc.pid ?? 0;
  } catch (err) {
    hookLog("idle_watcher_launch_failed", { error: String(err).slice(0, 300) });
    return 0;
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// ─── CLI entry point ──────────────────────────────────────────────────────────

/**
 * When this module is run directly, parse the config JSON arg and run
 * the watcher loop.
 */
async function main(): Promise<void> {
  const args = process.argv.slice(2);
  if (args.length < 1) {
    hookLog("idle_watcher_fatal_missing_args");
    process.exit(1);
  }

  try {
    const config = JSON.parse(args[0]) as IdleWatcherConfig;
    // Merge in any config overrides from the bootstrap.
    if (config.config) {
      const cfg = loadConfig();
      const overrides = config.config as Record<string, unknown>;
      if (overrides.base_url) {
        process.env.COGNEE_BASE_URL = String(overrides.base_url);
      }
      if (overrides.dataset) {
        config.dataset = String(overrides.dataset);
      }
    }
    await runIdleWatcher(config);
  } catch (err) {
    hookLog("idle_watcher_fatal_bad_args", { error: String(err).slice(0, 200) });
    process.exit(1);
  }
}

// Run main when this file is the entry point.
if (import.meta.main) {
  main();
}

export default runIdleWatcher;
